### PR TITLE
Python sdk 3.0.0

### PR DESCRIPTION
**Pull request by danielnaumau**

### CODE REVIEW

**Comment by danielnaumau** [comment]
All changes in the updated readme

**Comment by danielnaumau** [comment]
> For me this seems to be too simplified. Using this you loose all the flexibility our API has to offer.
> 
> This would be acceptable if all old functions were supported + we have all these new `simple versions` (like `sdk.simple.time_map(...)`) or them being present as some part of utils.
> 
> Coming from other languages this might be confusing. It might be harder to read documentation (since stuff is not inline). This implementation is also less performant since SDK has to map params into old structure.

All functions are supported. The only difference is that you can have one settings per one request

**Comment by danielnaumau** [comment]
Plus I took a look and people always use our sdk to create the same requests with the same params.

**Comment by danielnaumau** [comment]
I think it's worth to leave it that way and try to collect the feedback. Cause I think that all our sdks are quite uncomfortable for usage, especially for people who didn't see our api before. Plus I left a very detailed readme with all functions.

**Comment by danielnaumau** [comment]
> For me this seems to be too simplified. Using this you loose all the flexibility our API has to offer.
> 
> This would be acceptable if all old functions were supported + we have all these new `simple versions` (like `sdk.simple.time_map(...)`) or them being present as some part of utils.
> 
> Coming from other languages this might be confusing. It might be harder to read documentation (since stuff is not inline). This implementation is also less performant since SDK has to map params into old structure.

Plus I cannot do this way cause it will violate some python principles. We have too many duplicated names and in python libraries people import everything from one module.

**Comment by mjanuszkiewicz-tt** [comment]
> For me this seems to be too simplified. Using this you loose all the flexibility our API has to offer.
I think it's possbile to manually create and send a `TimeFilterRequest` object, it's just not mentioned in the readme?

**Comment by danielnaumau** [comment]
> > For me this seems to be too simplified. Using this you loose all the flexibility our API has to offer.
> 
> I think it's possbile to manually create and send a `TimeFilterRequest` object, it's just not mentioned in the readme?

Not in Python. People usually don't use types here and it will cause a lot of pain creating such complex nested classes

**Comment by danielnaumau** [comment]
Plus all python libraries have a very simplified import system. Everything should be imported from the same module. With our numerous ArrivalSearch and DepartureSearch, it will be very complicated to do smh like that.

File: README.md [hunk 1]
```diff
@@ -23,6 +23,22 @@ sdk = TravelTimeSdk('YOUR_APP_ID', 'YOUR_APP_KEY')
...
+sdk = TravelTimeSdk('YOUR_APP_ID', 'YOUR_APP_KEY')
+
+res = await sdk.time_map_async(
+    coordinates=[Coordinates(lat=51.507609, lng=-0.128315), Coordinates(lat=51.517609, lng=-0.138315)],
+    arrival_time=datetime.now(),
+    transportation=Driving()
+)
+```
+
+
```
Comment (Donatas-L) [inline README.md:41]: That's not a proper example, You need to show how to make 100x100 matrix by utilizing concurrency. Or something useful. Just dropping the async method ain't enough. All examples should be easy to understand and useful, not just code snippets which doesn't do anything.

File: traveltimepy/mapper.py [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+    transportation: Union[PublicTransport, Driving, Ferry, Walking, Cycling, DrivingTrain],
+    properties: Optional[List[Property]],
+    departure_time: Optional[datetime],
+    arrival_time: Optional[datetime],
+    travel_time: int,
+    full_range: Optional[FullRange]
+) -> TimeFilterRequest:
+    if properties is None:
+        properties = [Property.TRAVEL_TIME]
+    if arrival_time is not None:
```
Comment (mjanuszkiewicz-tt) [inline traveltimepy/mapper.py:44]: If both `arrival_time` and `departure_time` is set then we silently do an arrival search.

File: README.md (hunk 1 above)
Comment (danielnaumau) [reply README.md:41]: Look at benchmarks folder. There I have these examples.

File: traveltimepy/mapper.py (hunk 2 above)
Comment (danielnaumau) [reply traveltimepy/mapper.py:44]: yeah, good point, I'll add throwing errors there

**Comment by danielnaumau** [comment]
> > > > For me this seems to be too simplified. Using this you loose all the flexibility our API has to offer.
> > > 
> > > 
> > > I think it's possbile to manually create and send a `TimeFilterRequest` object, it's just not mentioned in the readme?
> > 
> > 
> > Not in Python. People usually don't use types here and it will cause a lot of pain creating such complex nested classes
> 
> Is there anything stopping a user from writing:
> 
> ```
> from traveltimepy.dto.requests.time_filter import TimeFilterRequest
> from traveltimepy.http import send_get
> 
> request = TimeFilterRequest(...)
> 
> result = send_post(
>   TimeFilterResponse,
>   'time-filter',
>   self.__headers(AcceptType.JSON),
>   request
> ).results
> ```
> 
> if they really wanted to?

Yeah, it's a bad practice in python. They don't use nested modules. Only smh like that: from traveltimepy import TimeFilterRequest

**Comment by mjanuszkiewicz-tt** [comment]
OK, but that's a social barrier which can be ignored if you need something done, not a technical one.

File: traveltimepy/accept_type.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from enum import Enum
+
+
+class AcceptType(Enum):
+    JSON = 'application/json'
+    BOUNDING_BOXES_JSON = 'application/vnd.bounding-boxes+json'
+    GEO_JSON = 'application/geo+json'
+    OCTET_STREAM = 'application/octet-stream'
```
Comment (MockusTravelTime) [inline traveltimepy/accept_type.py:8]: If this is supposed to reflect [this](https://docs.traveltime.com/api/reference/isochrones#Response-Body)
then it's missing few options

Comment (danielnaumau) [reply traveltimepy/accept_type.py:8]: They are not done yet. It will be the second pr

File: traveltimepy/mapper.py [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+                )
+                for departure_id, arrival_ids in searches.items()
+            ],
+            arrival_searches=[]
+        )
+    else:
+        raise ApiError('arrival_time or departure_time should be specified')
+
+
+def create_proto_request(
```
Comment (MockusTravelTime) [inline traveltimepy/mapper.py:489]: Coming from other languages  I would prefer this file split into multiple route based one (Or something similar)
If for whatever reasons this is preferred way of doing things in python, you can ignore my comment

File: traveltimepy/accept_type.py (hunk 3 above)
Comment (MockusTravelTime) [reply traveltimepy/accept_type.py:8]: When they'll be done add example to readme on how to call time map with geojson option too

File: traveltimepy/mapper.py (hunk 4 above)
Comment (danielnaumau) [reply traveltimepy/mapper.py:515]: yeah, I'll think about it. I think that I can make it better.

File: traveltimepy/accept_type.py (hunk 3 above)
Comment (danielnaumau) [reply traveltimepy/accept_type.py:8]: yeah, of course. This pr is huge, so I didn't want to make it bigger.

**Pull request closed without merging**